aws/src/main/java/org/apache/iceberg/aws/dynamodb/DynamoDbCatalog.java

```diff
@@ -0,0 +1,613 @@
+    } else {
+      return CatalogUtil.loadFileIO(fileIOImpl, properties, hadoopConf);
+    }
+  }
+
+  private String cleanWarehousePath(String path) {
+    Preconditions.checkArgument(path != null && path.length() > 0,
+        "Cannot initialize DynamoDbCatalog because warehousePath must not be null");
+    int len = path.length();
+    if (path.charAt(len - 1) == '/') {
+      return path.substring(0, len - 1);
+    } else {
+      return path;
+    }
+  }
+
+  private void validateNamespace(Namespace namespace) {
+    for (String level : namespace.levels()) {
+      ValidationException.check(level != null && !level.isEmpty(),
+          "Namespace level must not be empty: %s", namespace);
+      ValidationException.check(!level.contains("."),
+          "Namespace level must not contain dot, but found %s in %s", level, namespace);
+    }
+  }
+
+  private void validateTableIdentifier(TableIdentifier identifier) {
+    validateNamespace(identifier.namespace());
+    ValidationException.check(identifier.hasNamespace(),
+        "Table namespace must not be empty: %s", identifier);
+    String tableName = identifier.name();
+    ValidationException.check(!tableName.contains("."),
+        "Table name must not contain dot: %s", tableName);
+  }
+
+  private boolean dynamoDbTableExists(String tableName) {
+    try {
+      dynamo.describeTable(DescribeTableRequest.builder()
+          .tableName(tableName)
+          .build());
+      return true;
+    } catch (ResourceNotFoundException e) {
+      return false;
+    }
+  }
+
+  private void ensureCatalogTableExistsOrCreate() {
+
+    if (dynamoDbTableExists(awsProperties.dynamoDbTableName())) {
+      return;
+    }
+
+    LOG.info("DynamoDb catalog table {} not found, trying to create", awsProperties.dynamoDbTableName());
+    dynamo.createTable(CreateTableRequest.builder()
+        .tableName(awsProperties.dynamoDbTableName())
+        .keySchema(
+            KeySchemaElement.builder().attributeName(COL_IDENTIFIER).keyType(KeyType.HASH).build(),
+            KeySchemaElement.builder().attributeName(COL_NAMESPACE).keyType(KeyType.RANGE).build())
+        .attributeDefinitions(
+            AttributeDefinition.builder().attributeName(COL_IDENTIFIER).attributeType(ScalarAttributeType.S).build(),
+            AttributeDefinition.builder().attributeName(COL_NAMESPACE).attributeType(ScalarAttributeType.S).build())
+        .globalSecondaryIndexes(GlobalSecondaryIndex.builder()
+            .indexName(GSI_NAMESPACE_IDENTIFIER)
+            .keySchema(
+                KeySchemaElement.builder().attributeName(COL_NAMESPACE).keyType(KeyType.HASH).build(),
+                KeySchemaElement.builder().attributeName(COL_IDENTIFIER).keyType(KeyType.RANGE).build())
+            .projection(Projection.builder().projectionType(ProjectionType.KEYS_ONLY).build())
+            .build())
+        .billingMode(BillingMode.PAY_PER_REQUEST)
+        .build());
+
+    Tasks.foreach(awsProperties.dynamoDbTableName())
```

Comment:
I assume what we're doing here is waiting for dynamodb to complete provisioning of the table (not sure how long that typically takes).  It might be good to add a comment to explain why this check is necessary for future reference.

Reply:
It usually takes just a few seconds. I will add some comments.